daliuge-translator/dlg/dropmake/web/main.js

```diff
@@ -312,93 +379,88 @@ function fillOutSettings() {
     }
 
     //fill out settings list rom deploy methods array
-    var deployMethodManagerDiv = $("#DeployMethodManager")
+    const deployMethodManagerDiv = $("#DeployMethodManager");
     deployMethodManagerDiv.empty()
     console.log("filling out settings, GET Errors and Cors warning from Url check")
     deployMethodsArray.forEach(async element => {
+        let i;
+        const urlReachable = await checkUrlStatus(element.url);
+        const directlyAvailableMethods = await checkUrlSubmissionMethods(element.url + '/api/submission_method');
+        const translatorAvailableMethods = await checkUrlSubmissionMethods(getCurrentPageUrl() + '/api/submission_method?dlg_mgr_url=' + element.url);
+        let ReachableIcon;
 
-        var urlReachable = await checkUrlStatus(element.url)
-        var ReachableIcon = ""
-        
-        if(urlReachable){
+        if (urlReachable) {
             ReachableIcon = `<div class="settingsInputTooltip tooltip tooltipBottom urlStatusIcon urlStatusReachable" data-text="Destination URL Is Reachable, click to re-check" onclick="manualCheckUrlStatus('icon')"><i class="material-icons md-24">done</i></div>`
-        }else{
+        } else {
             ReachableIcon = `<div class="settingsInputTooltip tooltip tooltipBottom urlStatusIcon urlStatusNotReachable" data-text="Destination URL Is Not Reachable, click to re-check" onclick="manualCheckUrlStatus('icon')"><i class="material-icons md-24">close</i></div>`
         }
 
-        var directOption =  '<option value="direct">Direct</option>'
-        var helmOption =  '<option value="helm">Helm</option>'
-        var restOODOption =  '<option value="rest-ood">Rest-OOD</option>'
-        var restDirectOption = '<option value="rest-direct">Rest-Direct</option>'
-
-        if(element.deployMethod === "direct"){
-            directOption =  '<option value="direct" selected="true">Direct</option>'
-        }else if(element.deployMethod === "helm"){
-            helmOption =  '<option value="helm" selected="true">Helm</option>'
-        }else if(element.deployMethod === "rest-ood"){
-            restOODOption = '<option value="rest-ood" selected="true">Rest-OOD</option>'
-        } else if(element.deployMethod === "rest-direct"){
-            restDirectOption = '<option value="rest-direct" selected="true">Rest-Direct</option>'
+        const allAvailableMethods = directlyAvailableMethods["methods"].concat(translatorAvailableMethods["methods"]);
+        const availableOptions = [];
+        const default_options = ["SERVER", "BROWSER", "HELM", "OOD"];
+        if (allAvailableMethods.length === 0){  // Support backend without submission/method api
+            default_options.forEach((option, i) => availableOptions.push(buildDeployMethodEntry(option, i === 0)))
+        } else {
+            for (i = 0; i < allAvailableMethods.length; i++) {
+                const deploy_option = allAvailableMethods[i];
+                if (element.deployMethod !== "undefined") {
```

Comment:
**issue**: Checking against a string `undefined` smells to me (or maybe is just wrong). Is this really how the deploy methods are stored in localstorage?

Reply:
I think JS coercion makes this unclear; I've changed this to simply `undefined`.
I should probably reverse this conditional to make things clearer.
I'm catching the case where the value *is* set first.